pkg/trace/api/api.go

```diff
@@ -216,62 +215,81 @@ func (r *HTTPReceiver) handleTraces(v Version, w http.ResponseWriter, req *http.
 		return
 	}
 
-	traces, ok := getTraces(v, w, req)
-	if !ok {
-		return
-	}
-
-	// We successfully decoded the payload
-	r.replyTraces(v, w)
-
-	// We parse the tags from the header
-	tags := info.Tags{
+	ts := r.Stats.GetTagStats(info.Tags{
 		Lang:          req.Header.Get("Datadog-Meta-Lang"),
 		LangVersion:   req.Header.Get("Datadog-Meta-Lang-Version"),
 		Interpreter:   req.Header.Get("Datadog-Meta-Lang-Interpreter"),
 		TracerVersion: req.Header.Get("Datadog-Meta-Tracer-Version"),
-	}
+	})
 
-	// We get the address of the struct holding the stats associated to the tags
-	ts := r.Stats.GetTagStats(tags)
+	mediaType := getMediaType(req)
+	switch mediaType {
+	case "application/msgpack":
+		// process sequentially as we decode
+		for trace := range decodeTraces(w, req, v) {
+			r.sendTrace(ts, trace)
+		}
+	default:
+		traces, ok := getTraces(v, w, req)
+		if !ok {
+			return
+		}
+		for _, trace := range traces {
+			r.sendTrace(ts, trace)
+		}
+	}
 
 	bytesRead := req.Body.(*LimitedReader).Count
 	if bytesRead > 0 {
 		atomic.AddInt64(&ts.TracesBytes, int64(bytesRead))
 	}
+	r.replyTraces(v, w)
```

Comment:
Note that one change here is that we're only replying to the clients _after_ we're done processing. Previously were replying as soon as we decoded. This could be a good improvement because we could extend further and reply with any potential normalization errors to clients at the cost of longer lasting HTTP transactions.

Reply:
Since we actually don't reply anything different from the processing, any reason to move it? 
I don't know the full impact of holding tracer's connections for longer. There is also a risk to reach tracer's timeout?
Because of that, if there is no value, I'd rather not take the risk.

Reply:
I agree. I moved it back in the latest commit. For `application/msgpack` it wouldn't have been much different since we decode in the for loop and simply send down the channel (which is async and doesn't block if it's full - we just drop) so not much change there. The change would've been bigger for JSON.